### ISSUE TITLE
Unable to do round trip IO with GPKG files in geopandas

### ISSUE THREAD

**Issue by pmlpm1986** [Toblerity/Fiona#1270]
## Expected behavior and actual behavior.

I have using GeoDataFrame objects in a project that involves IO and the GPKG format and I have come across a potential bug. geopandas uses fiona for IO, so I thought this would be the proper forum.

Previously (with Fiona 1.8), I was able to write a file with a GeoDataFrame column of Nones and read it back as Nones too.

Now (with Fiona 1.9.3), I read that file as NaN, which requires me to do pre-processing before writing the file. The previous behaviour makes more sense to me but perhaps there is a reason for this change.

Is this a bug or was something changed intentionally?

## Steps to reproduce the problem.

The issue can be reproduced as follows:

```
from geopandas import GeoDataFrame, read_file
from shapely.geometry import Point
gdf = GeoDataFrame(
    {
      'data': [None,None],
      'new': [3.4,4.5], 
      'geometry': [Point(1, 2), Point(3,4)],
      },
    index=['a','b']
    )
gdf.to_file('test.gpkg')
gdf2 = read_file('test.gpkg')
print(gdf)
print(gdf2)
```

And the output is:

```console
   data  new                 geometry
a  None  3.4  POINT (1.00000 2.00000)
b  None  4.5  POINT (3.00000 4.00000)
  index  data  new                 geometry
0     a   NaN  3.4  POINT (1.00000 2.00000)
1     b   NaN  4.5  POINT (3.00000 4.00000)
```

## Operating system

OS: Debian Bookworm
Python: 3.11.2

## Fiona and GDAL version and provenance

Fiona: 1.9.3 (via pip 23.0.1)
geopandas: 0.13.0

**Comment by jorisvandenbossche** [comment]
Which version of pandas are you using? (and were using previously together with fiona 1.8) 
(This might also be a change in behaviour in the (geo)dataframe constructor)

**Comment by pmlpm1986** [comment]
A colleague was able to reproduce the same issue just now using the code above with pandas 1.5.x, fiona 1.9.4, geopandas 0.13 and python 3.10 on linux.

**Comment by sgillies** [comment]
@pmlpm1986 I'm skeptical that the problem is in this project. Can you provide a reproducing script that doesn't involve geopandas?

**Comment by jorisvandenbossche** [comment]
Yes, I don't think this is related to fiona. Reading the same file (written with the example code above using fiona 1.9) with fiona 1.8 and 1.9, I see the same results:

```python
>>> fiona.__version__
'1.9.0'

>>> with fiona.open("test.gpkg") as col:
...     print(col.schema)
...     print([list(feat["properties"].items()) for feat in col])
{'properties': OrderedDict([('index', 'str'), ('data', 'str'), ('new', 'float')]), 'geometry': 'Point'}
[[('index', 'a'), ('data', None), ('new', 3.4)], [('index', 'b'), ('data', None), ('new', 4.5)]]
```


```python
>>> fiona.__version__
'1.8.18'

>>> with fiona.open("test.gpkg") as col:
...     print(col.schema)
...     print([list(feat["properties"].items()) for feat in col])
{'properties': OrderedDict([('index', 'str'), ('data', 'str'), ('new', 'float')]), 'geometry': 'Point'}
[[('index', 'a'), ('data', None), ('new', 3.4)], [('index', 'b'), ('data', None), ('new', 4.5)]]
```

Returning `None` values for the "data" column in both cases.

**Comment by jorisvandenbossche** [comment]
Actually, there is _is_ something that changed in fiona, and this has a side effect in the geopandas code. Because after my comment above, I was looking into the geopandas constructor and this actually didn't change behaviour. But because of how `GeoDataFrame.from_features` is implemented, it now _receives_ different data: the `from_features` method will, when looping over the features, check for each feature whether it has a `__geo_interface__` attribute, and if so use this. 

And this was added in fiona 1.9, I think, and it returns slightly different data than using the method above (directly accessing "properties" on the fiona feature):

```
In [5]: with fiona.open("test.gpkg") as col:
   ...:     print(col.schema)
   ...:     print([feat.__geo_interface__["properties"] for feat in col])
{'properties': OrderedDict([('index', 'str'), ('data', 'str'), ('new', 'float')]), 'geometry': 'Point'}
[{'index': 'a', 'new': 3.4}, {'index': 'b', 'new': 4.5}]
```

So in this case, when a certain feature's property value is "null", it is omitted from the "properties" dictionary, instead of being included as `None`. 

And this slight difference in behaviour between `feat["properties"]` and `feat.__geo_interface__["properties"]` is what causes the different result.

Now, I don't know if this was an intentional behaviour difference. And if so, then the new behaviour in geopandas.read_file is a consequence of that, and probably not something we want to workaround to keep the original behaviour (this will only matter if _all_ values of a column are null, I think)

**Comment by martinfleis** [comment]
I would actually assume that the key with the missing value is part of the `__geo_interface__` in this case, not that it is omitted. I was looking at the report on GeoPandas side and have a tendency to say that this should be fixed here, not there.

**Comment by sgillies** [comment]
Filtering out null properties wasn't intended. I've got a fix for this in #1276.

**Comment by sgillies** [comment]
Resolved by #1276.
